haystack/core/pipeline/utils.py

```diff
@@ -18,3 +20,103 @@ def parse_connect_string(connection: str) -> Tuple[str, Optional[str]]:
         split_str = connection.split(".", maxsplit=1)
         return (split_str[0], split_str[1])
     return connection, None
+
+
+class FIFOPriorityQueue:
+    """
+    A priority queue that maintains FIFO order for items of equal priority.
+
+    Items with the same priority are processed in the order they were added.
+    This queue ensures that when multiple items share the same priority level,
+    they are dequeued in the same order they were enqueued (First-In-First-Out).
+    """
+
+    def __init__(self) -> None:
+        """
+        Initialize a new FIFO priority queue.
+        """
+        # List of tuples (priority, count, item) where count ensures FIFO order
+        self._queue: List[Tuple[int, int, Any]] = []
+        # Counter to maintain insertion order for equal priorities
+        self._counter = count()
+
+    def push(self, item: Any, priority: int) -> None:
+        """
+        Push an item into the queue with a given priority.
+
+        Items with equal priority maintain FIFO ordering based on insertion time.
+        Lower priority numbers are dequeued first.
+
+        :param item:
+            The item to insert into the queue.
+        :param priority:
+            Priority level for the item. Lower numbers indicate higher priority.
+        """
+        count = next(self._counter)
```

Comment:
although within a different scope, this variable `count` has the same name as `from itertools import count`
maybe a different name like `order` - just to be safe